### ISSUE TITLE
Changing the behavior of WhenTypeIs for value types

### ISSUE THREAD

**Issue by jnyrup** [fluentassertions/fluentassertions#799]
Currently `WhenTypeIs` checks the runtime type of the expectation.
This can cause a `NullReferenceException` in `CreateFromEquivalencyValidationContext` in the following example.

```c#
var obj1 = new { date = (DateTime?)null };
var obj2 = new { date = (DateTime?)default(DateTime) };
var actual = new[] { obj1, obj2 };
var expected = new[] { obj2, obj1 };

actual.Should().BeEquivalentTo(expected, options => options
	.Using<DateTime>(ctx => ctx.Subject.Should().BeCloseTo(ctx.Expectation, 1.Seconds()))
	.WhenTypeIs<DateTime>());
```

As the runtime type of `default(DateTime)` is `DateTime`, `WhenTypeIs` will match.
`CreateFromEquivalencyValidationContext` will then cast the `null` subject to a `DateTime` causing a `NullReferenceException`.

Another case is when a struct implements an interface, and `CreateFromEquivalencyValidationContext` cast a null into the struct type.

I propose that we change the logic of `WhenTypeIs` to consider whether the `RuntimeType` or `CompileTimeType` is a value type.

I have implemented the proposal [here](
https://github.com/jnyrup/fluentassertions/commit/cd1227f52bee262933a3a98b5b5b893baa4eee4).
It is not a compile-breaking change, but if you relied on `WhenTypeIs<DateTime>` to match `DateTime?` it will break your code.
As there aren't *that* many tests of `WhenTypeIs` I wanted to start a discussion on the intended/expected behavior before submitting a PR:

This was inspired by the problem presented in #778 .

**Comment by dennisdoomen** [comment]
So the real question becomes what we expect `WhenTypeIs` to do. A [recent change](https://github.com/fluentassertions/fluentassertions/commit/27d9e0d6cef51f82bf3a59a82f224d60d75d64f4) I did ensured that `BeEquivalentTo` will treat nullable enumerables as collections.

**Comment by JoeMoggridge** [comment]
I've just run into this issue in our tests, when upgrading to 4.2.2 -> 5.9.0. Caused me many days of confusion. I wonder whether it might be worth putting something in the release notes of 5.0 to warn about this breaking change?

I've now implemented the workaround suggested in #778. 

Looking forward to the fix in version 6.0.
